### PR TITLE
refactor: calculate constraints all at once

### PR DESCRIPTION
**Pull request by dmuldrew**
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose

This PR refactors the constraints function implementation to reduce the percentage of overall runtime from 48% to about 10%. This change appears to result in a 30-40% improvement in overall runtime. This PR also moves the call to `linprog` into the main loop which reduces the optimization code from about 20% of the overall runtime to 8%. These two changes together  result in a roughly 50% overall improvement in the runtime of the algorithm.

I also created two small testing datasets in the PR which consist of data for 5 cars:
https://github.com/Breakthrough-Energy/PreREISE/blob/dmuldrew/refactor_calculate_constraints/prereise/gather/demanddata/transportation_electrification/tests/test_census_data.csv 
and 30 cars:
https://github.com/Breakthrough-Energy/PreREISE/blob/dmuldrew/refactor_calculate_constraints/prereise/gather/demanddata/transportation_electrification/tests/profiling_census_data.csv 

### What the code is doing

Calculates constraints over the entire dataframe instead of calculating constraints for individual trip data within a double loop. Cost is still calculated within the loops since it is dependent on previous iterations.

### Testing

manual performance testing and automated integration testing

### Where to look

The performance-related changes are in smart_charging.py.

### Time estimate

~15min

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -26,3 +26,4 @@ powersimdata = "~=0.4.4"
 nrel-pysam = "~=2.2"
 pyproj = "~=3.0"
 pygrib = "~=2.1.4"
+line-profiler = "*"
```
Comment (rouille) [inline Pipfile]: Do we want to add this to the Pipfile. It seems that it could be installed by the user in his environment if needed.

File: prereise/gather/demanddata/transportation_electrification/const.py [hunk 2]
```diff
@@ -36,4 +36,5 @@
 
 safety_coefficient = 1
 
+# ???
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/const.py]: What is that?

File: prereise/gather/demanddata/transportation_electrification/smart_charging.py [hunk 3]
```diff
@@ -55,7 +55,7 @@
 ]
 
 
-def get_constraints(individual, kwhmi, power, trip_strategy, location_strategy, cost):
+def get_constraints(individual, kwhmi, power, trip_strategy, location_strategy):
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/smart_charging.py]: `cost` needs to be removed from docstring

File: Pipfile (hunk 1 above)
Comment (dmuldrew) [reply Pipfile]: I think we can potentially take it out as a dependency, we mainly just need it for profiling

File: prereise/gather/demanddata/transportation_electrification/smart_charging.py [hunk 4]
```diff
@@ -70,7 +70,8 @@ def get_constraints(individual, kwhmi, power, trip_strategy, location_strategy,
     :return: (*pandas.DataFrame*) -- a DataFrame adding the calculated constraints
         to an individual vehicle's data
     """
-    constraints_df = individual.copy()
+    constraints_df = individual
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/smart_charging.py]: Is that not a good thing to copy? Just want to make sure it does not matter if `individual` is modified

File: prereise/gather/demanddata/transportation_electrification/const.py (hunk 2 above)
Comment (dmuldrew) [reply prereise/gather/demanddata/transportation_electrification/const.py]: Just a note to myself that we need to ask what the `ER` variable is?

File: prereise/gather/demanddata/transportation_electrification/smart_charging.py (hunk 4 above)
Comment (dmuldrew) [reply prereise/gather/demanddata/transportation_electrification/smart_charging.py]: I think unnecessary copying might be slowing things down in this code, I think it might potentially more performant to use views:
https://www.skytowner.com/explore/difference_between_copy_and_view_in_pandas

Comment (rouille) [reply prereise/gather/demanddata/transportation_electrification/smart_charging.py]: Just saw in the `smart_charging` function that `individual` is updated in each iteration and hence no copy is needed

File: prereise/gather/demanddata/transportation_electrification/tests/test_smart_charging.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+def test_smart_charging():
+    run_smart_charging()
+
+
+def run_smart_charging():
+    LOAD = loadmat(
+        "/workspaces/PreREISE/prereise/gather/demanddata/transportation_electrification/CAISO_sample_load_2019.mat"
+    )["load_demand"]
+    print(LOAD)
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/tests/test_smart_charging.py]: Not sure we want to keep the `print`

File: prereise/gather/demanddata/transportation_electrification/tests/test_smart_charging.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        power=6.6,
+        location_strategy=2,
+        veh_type="LDV",
+        filepath="/workspaces/PreREISE/prereise/gather/demanddata/transportation_electrification/nhts_census_updated_dwell.mat",
+        daily_values=daily_values,
+        LOAD=LOAD,
+        trip_strategy=1,
+    )
+
+    print(result)
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/tests/test_smart_charging.py]: Same here

File: prereise/gather/demanddata/transportation_electrification/smart_charging.py [hunk 7]
```diff
@@ -70,7 +70,8 @@ def get_constraints(individual, kwhmi, power, trip_strategy, location_strategy,
...
+    # constraints_df = individual.copy()
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/smart_charging.py]: I think you can remove this line

File: prereise/gather/demanddata/transportation_electrification/smart_charging.py [hunk 8]
```diff
@@ -385,7 +387,7 @@ def smart_charging(
                     if exitflag == 0:
 
                         # can be an EV
-                        individual = individual.copy()
+                        # individual = individual.copy()
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/smart_charging.py]: You can remove this line too

**Comment by rouille** [comment]
It is out of the scope of this PR but why `segsum` and `segcum` are parameters of the `calculate_optimization` function since these could be derive from `seg` within the function?

See line 365 of the `smart_charging` module
```
segsum = sum(seg)
segcum = np.cumsum(seg)
linprog_result = calculate_optimization(
    charging_consumption,
    rates,
    elimit,
    seg,
    segsum,
    segcum,
    total_trips,
    kwh,
)
```

**Comment by rouille** [comment]
Other comment out of the scope of this PR. In the `smart_charging` function. the `kwhmi` parameter is said to vary with `model_year` and `veh_type` that are also parameters of the function. can we have a dictionary that would map `(model_year, veh_type)` to `kwhmi` to simplify the calling of the function and also make it less obscure to the user. In your test, `kwhmi=0.242`, who would come up with that value.

**Comment by dmuldrew** [comment]
@rouille Yeah, I agree a dictionary is a good idea…we’d probably need to request that from them so we can have more comprehensive tests of the various code pathways.

**Comment by dmuldrew** [comment]
Latest changes reduce the optimization from around ~20% to ~8% of total runtime:
``` python
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   370     14166    3624525.0    255.9      1.8                      linprog_inputs = calculate_optimization(
   371      7083      14128.0      2.0      0.0                          charging_consumption,
   372      7083      14369.0      2.0      0.0                          rates,
   373      7083      14190.0      2.0      0.0                          elimit,
   374      7083      14488.0      2.0      0.0                          seg,
   375      7083      14036.0      2.0      0.0                          segsum,
   376      7083      14784.0      2.1      0.0                          segcum,
   377      7083      13873.0      2.0      0.0                          total_trips,
   378      7083      14797.0      2.1      0.0                          kwh,
   379                                                               )
   380                                           
   381      7083   12041961.0   1700.1      6.0                      linprog_result = linprog(**linprog_inputs)
```
so now this single line accounts for almost half the runtime (compared to 18% before):
``` python
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   493                                                                   # copy individual back to newdata if it can be an EV
   494      6804   94103836.0  13830.7     47.2                          newdata.iloc[i : i + total_trips] = individual
```

File: prereise/gather/demanddata/transportation_electrification/data_helper.py [hunk 9]
```diff
@@ -111,7 +117,13 @@ def generate_daily_weighting(year, area_type="urban"):
...
     if area_type not in allowable_area_types:
         raise ValueError(f"area_type must be one of {allowable_area_types}")
-    data_dir = os.path.join(os.path.dirname(__file__), "data")
+    data_dir = os.path.join(
+        os.path.dirname(inspect.getsourcefile(prereise)),
+        "gather",
+        "demanddata",
+        "transportation_electrification",
+        "data",
+    )
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/data_helper.py:126]: What is the purpose of the change? Just being curious.

File: prereise/gather/demanddata/transportation_electrification/smart_charging.py [hunk 10]
```diff
@@ -222,14 +213,14 @@ def calculate_optimization(
...
-        bounds=bounds,
-    )
+    return {
+        "c": f,
+        "A_ub": Aineq,
+        "b_ub": Bineq,
+        "A_eq": Aeq,
+        "b_eq": Beq,
+        "bounds": bounds,
+    }
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/smart_charging.py:221]: docstring needs to be updated accordingly.

File: prereise/gather/demanddata/transportation_electrification/tests/test_smart_charging.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        location_strategy=2,
+        veh_type="LDV",
+        filepath=os.path.join(
+            os.path.dirname(inspect.getsourcefile(prereise)),
+            "gather",
+            "demanddata",
+            "transportation_electrification",
+            "nhts_census_updated_dwell",
+        ),
+    )
```
Comment (rouille) [inline prereise/gather/demanddata/transportation_electrification/tests/test_smart_charging.py:35]: What are we testing here?

Comment (dmuldrew) [reply prereise/gather/demanddata/transportation_electrification/tests/test_smart_charging.py:35]: At this point, whether the code runs successfully without error

**Review by rouille — APPROVED** [review]
Looks good. Thanks

File: prereise/gather/demanddata/transportation_electrification/data_helper.py (hunk 9 above)
Comment (dmuldrew) [reply prereise/gather/demanddata/transportation_electrification/data_helper.py:126]: Just consistent with the method I used on ECTF, I had some issues with file paths on these files
